### PR TITLE
Add support for OpenJ9's gc policies

### PR DESCRIPTION
**Pull request by DanHeidinga**
Add support for OpenJ9's GC policies: gencon, balanced, opthruput, optavgpause, & metronome.

Map the pool names that can be returned by the MxBeans as either YOUNG or OLD.

Update tests to support OpenJ9 as well.

I've tested this locally on my mac using each of the policies except metronome (that policy isn't supported on osx).

Fixes: #1458

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@DanHeidinga Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=2514)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=2514) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by DanHeidinga** [comment]
I'm covered under a corporate CLA but can't seem to pass the CLA check.  Any advice on how to solve this?

**Comment by jonatan-ivanov** [comment]
@DanHeidinga Thanks for the PR. We are checking this and get back to you.

**Comment by DanHeidinga** [comment]
Thanks for looking into this.

> 1. Check the [FAQ](https://cla.pivotal.io/about#corporate-signature-email) and add your company email to your GH account (after this, you need to sync the status of this PR manually, please see the comment of the CLA bot)

This is the case I'm under - my Github account lists my company email as my primary email.  Also all the commits in this PR are signed with my company email.  I've validated this a couple of times and tried the manual sync as well to no effect.


> 2\. Check the [FAQ](https://cla.pivotal.io/about#corporate-signature-github-org) and make sure you are a member of your company's GH org and your membership is public
>     (after this, you need to sync the status of this PR manually, please see the comment of the CLA bot)

I'm not aware of a single GH org for my company so I haven't done this.

**Review by jonatan-ivanov — APPROVED** [review]

**Comment by jonatan-ivanov** [comment]
@DanHeidinga I saw that the committer email is matching, though I'm not sure if the CLA bot cares about it, I think it checks the email address assigned to your account. I see your email address on your profile (sorry, I did not check earlier) so everything should be ok on your side. Unfortunately, I'm not very familiar with the CLA bot nor have I the access to it but I already reported the issue hopefully it will be resolved soon.

**Comment by jonatan-ivanov** [comment]
The PR looks good to me but we need to figure out how to run the OpenJ9-related Gradle tasks.

Right now CircleCI does not have an OpenJ9 image and if this won't change in the near future we either need to build and maintain one for ourselves or move to a different CI solution that supports OpenJ9 out of the box.
I created a in issue for CircleCI to add support for it: https://github.com/CircleCI-Public/cimg-openjdk/issues/78

/cc @shakuzen

**Comment by pivotal-issuemaster** [comment]
@DanHeidinga Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=micrometer-metrics/micrometer&pullRequestId=2514)!

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java
Comment (shakuzen) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java:59]: I wasn't sure from reading [this documentation on the balanced collector](https://www.eclipse.org/openj9/docs/gc/#gc-processing_1) how the `balanced-reserved` region was used. I'm trying to figure out how we should calculate the live and max size. For comparison, with OpenJDK collectors we calculate it as the used/max size of the old gen pool after a GC, respectively.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java
Comment (shakuzen) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java:60]: The `optavgpause` and `optthruput` collectors are non-generational, but we can't detect that from the memory pool name alone since it uses `tenured-SOA` and `tenured-LOA` the same as gencon does. For generational collectors, we calculate the allocated bytes from the young (eden/nursery) pool, but for non-generational collectors we can't do that. Maybe we can rework the current logic somehow without too much special casing.

**Review by shakuzen — COMMENTED** [review]
I went through and took a closer look at things with each of the collectors (except metronome since I'm on OSX). I think we'll need some adjustments to get everything working. I've left more specific comments in the review.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java:50]: I'm updating the naming to make this more clear in #2554

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java:59]: I've opened #2562 to handle multiple long-lived heap pools. I'll rebase this pull request on that once it is merged.

Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java:59]: I added some less than ideal logic for detecting generational vs non in #2560. It works with the current implementation of all garbage collectors we claim to support even after this pull request, but it does seem fragile.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java
Comment (DanHeidinga) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java:269]: I had added it for completeness but you're completely right it won't issue any notifications.  Will remove

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java
Comment (DanHeidinga) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java:44]: That's correct.  I checked the code and got confirmation from the openj9 gc devs that this doesn't apply.  Will remove the comment

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java
Comment (DanHeidinga) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java:58]: I'm still trying to track this one down.  Will update when I get confirmation from the gc devs

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java:269]: I did notice some notifications for it when I was testing, but I'm not sure when they happen since they should be a no-op. Somehow it still reports some time taken for it's "GC event". We could consider supporting it if we understand what it means and there's interest from users.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java
Comment (DanHeidinga) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java:58]: `balanced-reserved` represents regions in the heap that are always empty.  They'll eventually be used as either eden or  tenure space.

The used space should always be 0 as they are always empty.  The max size should conceptually start at the  entire heap size as nothing will have been allocated yet.

Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java:58]: If I'm understanding correctly, then it sounds like we shouldn't use it in any of the calculations.

Comment (DanHeidinga) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmMemory.java:58]: I agree.  It might be interesting to track as its own thing but shouldn't be rolled up into any other calculation

**Review by shakuzen — APPROVED** [review]
Thanks for all the work on this. I added in metronome's memory pool, and I have tested locally (via docker).

**Pull request merged**
